src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+
+enum class FileFormat {
+	// Not yet detected or unrecognized
+	Unknown,
+	// Used by MS-DOS and most clones
+	MsDos,
+	// Used by Windows NT, slight variation of MS-DOS format
+	WinNt,
+	// Used by DR-DOS, encodes screen fonts in much more efficient way
+	DrDos
+};
+
+// Supported file size limit, just for some extra safety.
+// - Almost all the CPI files in the wild are below 64 KB size (MS DOS does not
+//   support anything bigger)
+// - 'DISPLAY.CPI' file from PTS-DOS 2000 is below 88 KB
+// - 'EGA.CPI' file from Windows 2000 (English) is slightly below 128 KB
+constexpr uint32_t MaxFileSize = 1024 * 1024;
+
+constexpr uint8_t IdMsDosWinNt = 0xff;
+constexpr uint8_t IdDrDos      = 0x7f;
+
+static const std::string SignatureMsDos = "FONT   ";
+static const std::string SignatureWinNt = "FONT.NT";
+static const std::string SignatureDrDos = "DRFONT ";
+
+static const std::string SignatureUpx = "UPX!";
+
+static const std::string DeviceEga = "EGA";
+static const std::string DeviceLcd = "LCD";
+static const std::set<std::string> KnownPrinterDevices =
+        {"4201", "4208", "5202", "1050", "EPS", "PPDS"};
+
+// Some CPI files (like 'DISPLAY.CPI' from PTS-DOS 2000 or '4208.CPI' from
+// MS-DOS 6.22) contain obviously dummy size or offset values in some places;
+// ignore them and don't spam the log output with warnings.
+static const std::set<uint16_t> DummySizeValues   = {0, 1, UINT16_MAX};
+static const std::set<uint32_t> DummyOffsetValues = {0, 1, UINT32_MAX};
+
+// ***************************************************************************
+// Storage data types
+// ***************************************************************************
+
+// Number of characters we need; the real font in the CPI file can contain a
+// different number of characters.
+constexpr uint16_t NumCharactersInFont = 256;
+
+struct ScreenFont {
+	static constexpr uint32_t FullSize_8x16 = NumCharactersInFont * 16;
+	static constexpr uint32_t FullSize_8x14 = NumCharactersInFont * 14;
+	static constexpr uint32_t FullSize_8x8  = NumCharactersInFont * 8;
+
+	std::vector<uint8_t> font_8x16 = {};
+	std::vector<uint8_t> font_8x14 = {};
+	std::vector<uint8_t> font_8x8  = {};
+};
+
+using ScreenFonts = std::unordered_map<uint16_t, ScreenFont>;
+
+// ***************************************************************************
+// Persistend data storage
+// ***************************************************************************
+
+static ScreenFonts ega_font_storage = {};
+
+// ***************************************************************************
+// CPI file structure reader
+// ***************************************************************************
+
+struct FontFileHeader {
+	static constexpr auto StructName = "FontFileHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint8_t id_byte       = 0;
+	std::string id_string = {};
+	uint16_t num_pointers = 0;
+	uint8_t pointer_type  = 0;
+	uint32_t fih_offset   = 0;
```

Comment:
What's `fih`? Probably an explanatory comment is in order. Better yet, consider adding explanatory comments to all struct members.

Reply:
Added comments, everywhere.